docs/client-design.md

```diff
@@ -0,0 +1,287 @@
+ * sockets, we provide a wrapper class around the `Data`, which should be passed 
+ * as the `ptr` argument of `epoll_data` to `epoll` by the application for its 
+ * own file descriptors.
+ */
+template<class Data *>
+class EpollIOEventProviderData {
+public:
+    EpollIOEventProviderData(int type, Data *data);
+    
+    /** 
+     * Type of file descriptor, needed for distinguishing application file 
+     * descriptors from tntcxx sockets. 
+     */
+    int type;
+    Data *data;
+};
+
+/**
+ * Encapsulates calling of notification callbacks from epoll for tntcxx. 
+ */
+template<class Data *>
+class EpollIoEventProviderDataTntcxx : public EpollIOEventProviderData<Data *> {
+public:
+    EpollIoEventProviderData(int fd, Data *data, read_ready_cb_f read_ready_cb, write_ready_cb_f write_ready_cb);
+    
+    /** Needs to be called by the application on a read event. */
+    void read_ready();
+    /** Needs to be called by the application on a write event. */
+    void write_ready();
+};
+```
+
+### Connections
+
+```c++
+/**
+ * A connection encapsulates sending requests and receiving responses from one 
+ * Tarantool instance.
+ */
+template<class IOEventProvider>
+class Connection {
+public:
+    Connection(IOEventProvider &net_provider, 
+	           const ConnectionOptions &connection_options);
+	
+    /** 
+	 * Return the connection's status.
+	 */
+	enum class ConnectionStatus get_status() const;
+	
+	/** 
+	 * If the connection is an erroneous state (see `status`), return the 
+	 * connection error that caused it. The same error is also passed to 
+	 * response callbacks, and the same error is returned by 
+	 * `Request::get_error`. 
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/* An abstract request's interface. A request object is always returned. */
+	Request some_request(/* options */);
+};
+```
+
+#### Connect Options
+
+```c++
+/* Extend `ConnectionOptions` with an option for the reconnection feature. */
+struct ConnectOptions {
+	/* All existing options. */
+	
+	/** 
+	 * In the event of a broken connection, the interval in which the stream 
+	 * tries to re-establish the connection. 
+	 */
+	static constexpr size_t DEFAULT_RECONNECTION_INTERVAL = 2;
+	size_t reconnection_interval = DEFAULT_RECONNECTION_INTERVAL;
+};
+```
+
+### Requests
+
+```c++
+/** Encapsulates management of a request issued through a connection. */
+class Request {
+public:
+	/** 
+	 * Callback called when response is ready. Since the callback has a fixed 
+	 * signature, we need to allow for capturing additional context using 
+	 * lambdas. Hence, we use `std::function` for type erasure.
+	 * 
+	 * See `Connection::get_error` for details about the `error` parameter.
+	 */ 
+	using request_cb_f = 
+		std::function<void(const std::optional<ConnectionError> &error, 
+		                   Response &&response, Connection &connection)>;
+	
+	/** Set a callback called when the response is ready. */
+	int set_callback(request_cb_f request_cb);
```

Comment:
Why does it return `int`?

Reply:
I was thinking that `set_callback` can fail, and there it no other approach to returning errors in this project. 

Let's make it `void`.